src/_bulk_response.ts

```diff
@@ -27,60 +27,87 @@ interface ByteModel {
 export const MULTI_QUERY_MAGIC_STR = "chal1";
 export const MULTI_QUERY_RESPONSE_MAGIC_STR = "CHALK_BYTE_TRANSMISSION";
 
-export function parseByteModel(raw: Buffer): ByteModel {
-  const decoder = new TextDecoder();
-
-  let offset = 0;
-
-  const responseMagicBytes = raw.subarray(
+const assertBufferBeginsWithMagicString = (
+  buf: Buffer,
+  decoder: TextDecoder,
+  offset = 0
+): number => {
+  const responseMagicBytes = buf.subarray(
     offset,
     offset + MULTI_QUERY_RESPONSE_MAGIC_STR.length
   );
-  offset += MULTI_QUERY_RESPONSE_MAGIC_STR.length;
   if (decoder.decode(responseMagicBytes) !== MULTI_QUERY_RESPONSE_MAGIC_STR) {
     throw new Error(
       "Bulk query response was not prefixed with appropriate string constant"
     );
   }
 
-  let curLen = Number(raw.readBigUInt64BE(offset));
-  offset += 8;
-  const attrsBytes = raw.subarray(offset, curLen + offset);
-  offset += curLen;
+  return offset + +MULTI_QUERY_RESPONSE_MAGIC_STR.length;
+};
+
+const readBufferFromOffset = (
+  buf: Buffer,
+  offset: number
+): [subBuffer: Buffer, newOffset: number] => {
+  const bufLength = Number(buf.readBigUInt64BE(offset));
+  const offsetFromContent = offset + 8;
+  const subBuffer = buf.subarray(
+    offsetFromContent,
+    offsetFromContent + bufLength
+  );
 
-  curLen = Number(raw.readBigUint64BE(offset));
-  offset += 8;
-  const pydanticBytes = raw.subarray(offset, curLen + offset);
-  offset += curLen;
+  return [subBuffer, offsetFromContent + bufLength];
+};
 
-  curLen = Number(raw.readBigUint64BE(offset));
-  offset += 8;
-  const jsonStr = decoder.decode(raw.subarray(offset, curLen + offset));
-  const jsonParsed = JSON.parse(jsonStr);
-  offset += curLen;
+export function parseByteModel(raw: Buffer): ByteModel {
```

Comment:
refactor for readability and some cleanup of repeated logic. This is already under unit tests :)